nav2_util/include/nav2_util/robot_utils.hpp

```diff
@@ -17,60 +18,24 @@
 #define NAV2_UTIL__ROBOT_UTILS_HPP_
 
 #include <string>
-#include <shared_mutex>
+#include <memory>
+#include <algorithm>
 #include <mutex>
-#include <rclcpp/rclcpp.hpp>
-#include <geometry_msgs/msg/pose_with_covariance_stamped.hpp>
-#include <geometry_msgs/msg/twist.hpp>
-#include <nav_msgs/msg/odometry.hpp>
+#include <thread>
 
-namespace nav2_util
-{
-
-class VelocityPublisher
-{
-public:
-  VelocityPublisher(
-    rclcpp::Node::SharedPtr & node,
-    const std::string topic = std::string("/cmd_vel"));
-
-  void publishCommand(const geometry_msgs::msg::Twist & cmd_vel);
-
-private:
-  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr vel_pub_;
-};
+#include <geometry_msgs/msg/pose_stamped.hpp>
+#include "tf2_ros/buffer.h"
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+#include "rclcpp/rclcpp.hpp"
 
-class RobotStateHelper
+namespace nav2_util
 {
-public:
-  RobotStateHelper(
-    rclcpp::Node::SharedPtr & node,
-    const std::string odom_topic = std::string("/odom"));
-
-  bool getOdometry(nav_msgs::msg::Odometry::SharedPtr & robot_odom);
-
-  bool getCurrentPose(
-    geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr & robot_pose,
-    const bool use_topic = true);
-
-private:
-  void onPoseReceived(geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr robot_pose);
-  void onOdomReceived(nav_msgs::msg::Odometry::SharedPtr msg);
-
-  bool getGlobalLocalizerPose(
-    geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr & robot_pose);
-  bool getTfPose(geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr & robot_pose);
 
-private:
-  rclcpp::Node::SharedPtr node_;
-  rclcpp::Subscription<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr pose_sub_;
-  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
-  std::shared_timed_mutex state_mutex_;
-  bool initial_pose_received_, initial_odom_received_;
-  geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr current_pose_;
-  nav_msgs::msg::Odometry::SharedPtr current_odom_;
-};
+bool getCurrentPose(
+  geometry_msgs::msg::PoseStamped & global_pose,
+  std::shared_ptr<tf2_ros::Buffer> & tf_buffer, const std::string global_frame = "map",
```

Comment:
maybe not necessary to pass a shared_ptr, in this case a reference might be enough if the funcion is not managing the lifetime of the buffer

Reply:
Well the host is going to have it stored on their end as a shared pointer because its good to have 1 buffer a bunch of people share, I could pass as `const shared_ptr<> &` if that's better, but I'd like to avoid exposing the raw object when possible using smart pointers

Frankly I dont know the best practices on that, so if that's the right answer, let me know and I'll just do it

Reply:
@crdelsey what do you think? here are some [guidelines](https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/)

Reply:
Awesome, I'll review that tomorrow and get back to you with updates on all your ptr related questions.

Reply:
IMHO, a `const shared_ptr<> &` is fine in this case. Passing a `tf2_ros::Buffer &` would be fine as well. I think I would have preferred it in many cases, but the way ROS 2 is written, working with native references and pointers can be clunky in many cases. Better to go with the flow and use `shared_ptr`s when dealing with ROS 2 types.

In this particular case, most callers will have a `shared_ptr` to their `Buffer`.  Taking a `Buffer &` will force them to pass in `* my_buffer` everywhere rather than just `my_buffer`. I'd rather not force them to throw in the `*`